### PR TITLE
Add Support for Mongodb to CB data migration

### PR DESCRIPTION
**Pull request by prajwal-pai77**
Adds the UI for MongoDB to CB data migration.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -541,6 +543,11 @@
         "title": "Data Export",
         "category": "Couchbase"
       },
+      {
+        "command": "vscode-couchbase.tools.mdbExport",
```
Comment (lokesh-couchbase) [inline package.json:547]: Shouldn't we rename this to mdbImport?

Applicable everywhere.

File: package.json [hunk 2]
```diff
@@ -541,6 +543,11 @@
...
+        "title": "MongoDb Migrate",
```
Comment (lokesh-couchbase) [inline package.json:548]: Let's make it MongoDB (B Capital)

File: src/extension.ts [hunk 3]
```diff
@@ -65,6 +65,7 @@ import { clearDocumentFilter } from "./commands/documents/clearDocumentFilter";
 import { getClusterOverviewData } from "./util/OverviewClusterUtils/getOverviewClusterData";
 import { checkAndCreatePrimaryIndex } from "./commands/indexes/checkAndCreatePrimaryIndex";
 import { dataExport } from "./pages/Tools/DataExport/dataExport";
+import {mdbExport} from "./pages/Tools/MDBExport/MDBExport";
```
Comment (lokesh-couchbase) [inline src/extension.ts:68]: follow formatting

File: package.json (hunk 1 above)
Comment (lokesh-couchbase) [reply package.json:547]: Even mdbMigrate sounds better

File: src/handlers/handleCLIDownloader.ts [hunk 4]
```diff
@@ -99,6 +104,14 @@ class DependenciesDownloader {
                     OSUtil.MACOS_64
                 )
             );
+            map.set(
+                this.TOOL_MDB_MIGRATE,
+                this.getToolSpec(
+                    "https://intellij-plugin-dependencies.s3.us-east-2.amazonaws.com/cbmigrate/cbmigrate_0.0.1-beta_darwin_amd64.zip",
```
Comment (lokesh-couchbase) [inline src/handlers/handleCLIDownloader.ts:110]: Please double check with @shyam-cb before the final build

File: src/handlers/handleCLIDownloader.ts [hunk 5]
```diff
@@ -252,6 +313,37 @@ class DependenciesDownloader {
             logger.debug("CBShell is already installed");
             this.setToolActive(ToolStatus.AVAILABLE, shellPath, shell);
         }
+
+        const cbMigrate = downloads.get(this.TOOL_MDB_MIGRATE);
```
Comment (lokesh-couchbase) [inline src/handlers/handleCLIDownloader.ts:317]: Can we move all these in their respective functions?

Even do it with other tools, As it has become kinda hard t read and understand now

File: src/handlers/handleCLIDownloader.ts [hunk 6]
```diff
@@ -252,6 +313,37 @@ class DependenciesDownloader {
...
+        if (
+            cbMigrateStatus === ToolStatus.NOT_AVAILABLE &&
+            !this.isInstalled(
+                toolsPath,
+                cbMigrateDownloadsMap,
+                CBToolsType.CB_MIGRATE
+            )
+        ) {
+            // Avoiding 2 threads to install the same thing at the same time
+            logger.info("Downloading CB Migrate..");
```
Comment (lokesh-couchbase) [inline src/handlers/handleCLIDownloader.ts:340]: Use single full stop at the end. Or add multiple to other tools as well

File: src/tools/MDBToCBExport.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                }
+                cmd.push("--cb-username");
+                cmd.push(connection.username);
+                cmd.push("--cb-password");
+                cmd.push(password);
+                cmd.push("--cb-bucket");
+                cmd.push(cbBucket);
+                cmd.push("--cb-scope");
+                cmd.push(cbScope);
+                // Collection field is optional
```
Comment (lokesh-couchbase) [inline src/tools/MDBToCBExport.ts:55]: Do we need this?
if not let's remove the comment

File: src/pages/Tools/MDBExport/MDBExport.ts [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+            logger.info("Connected successfully to MongoDB");
+            const adminDb = client.db().admin();
+            logger.info("fetching list of databases");
+            const databasesList = await adminDb.listDatabases();
+            const databases = databasesList.databases.map((db) => db.name);
+            await client.close();
+
+            return databases;
+        } catch (error) {
+            let error_message =
```
Comment (lokesh-couchbase) [inline src/pages/Tools/MDBExport/MDBExport.ts:145]: use const

File: src/tools/MDBToCBExport.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                // cmd.push(colCB);
+                cmd.push("--cb-generate-key");
+                cmd.push("%_id%");
+
+                cmd.push("; \n");
+
+                // Run Command
+                const terminal: vscode.Terminal =
+                    vscode.window.createTerminal("MDBToCb");
+                let text = cmd.join(" ");
```
Comment (lokesh-couchbase) [inline src/tools/MDBToCBExport.ts:66]: If text is constant, use const

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 10]
```diff
@@ -0,0 +1,554 @@
+export const dataMdbMigrateWebview = async (buckets: string[]): Promise<string> => {
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:1]: can we remove data from the name?

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+    return /*html*/`
+    <!DOCTYPE html>
+    <html lang="en">
+       <head>
+          <meta charset="UTF-8">
+          <meta name="viewport" content="width=device-width, initial-scale=1.0">
+          <title>Data Migrate</title>
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:8]: MongoDB Data Migrate?

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+            }
+            
+            /* Style for form labels */
+            label {
+                display: block;
+                margin-bottom: 5px;
+                font-weight: bold;
+            }
+
+            .verboseLogContainer {
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:63]: remove unused css

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+                display: flex;
+                gap: 10px;
+            }
+            input[type="checkbox"] {
+                transform: scale(1.1); 
+            }
+
+            input[type="text"], input[type="number"]{
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:71]: remove input type number

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+               padding-right: 10px;
+             }
+             
+             .separator {
+               flex: 1;
+               border: none;
+               border-top: 1px solid var(--vscode-settings-sashBorder);
+             }
+            
+            #selectedFolder {
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:174]: same

Please check all the CSS and remove unwanted ones

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+                bottom: -2px; 
+            }
+
+            input[type="checkbox"] {
+                width: 16px; 
+                height: 16px; 
+                vertical-align: middle;
+            }
+
+
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:261]: Remove extra lines

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+            
+          </style>
+        </head>
+
+        <body>
+        <h2 class="heading">Mongodb to Couchbase Data Migration</h2>
+        <form action="#" method="post" id="dataMigrateForm">
+        <p>This tool utilizes the CLI in the background to migrate the data, Please refer to <a href="https://github.com/couchbaselabs/cbmigrate" target="_blank">this</a> for more info on how to use the CLI.</p>
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:269]: make it in multiple lines, this is increasing the code's width.

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+           <div class="separator-container">
+              <span class="separator-text">Source</span>
+              <div class="separator"></div>
+           </div>
+           <label for="mongodbconnectionstring" class="tooltip">MongoDB Connection URI:
+           <span class="tooltiptext">Specify the MongoDB Connection URI</span>
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:275]: Please add formatting using html language features

File: src/webViews/tools/dataMdbMigrate.webview.ts [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+            break;
+        case "vscode-couchbase.tools.mdbExport.formValidationError":
+            const error = message.error;
+            document.getElementById("validation-error").innerHTML =
+                message.error;
+            break;
+    }
+});
+
+$(document).ready(function () {
```
Comment (lokesh-couchbase) [inline src/webViews/tools/dataMdbMigrate.webview.ts:512]: I don't think we require this as we don't have advanced settings

File: src/tools/MDBToCBExport.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
+import { CBTools, Type } from "../util/DependencyDownloaderUtils/CBTool";
+import { getActiveConnection, getConnectionId } from "../util/connections";
+import { Constants } from "../util/constants";
+import * as keytar from "keytar";
+import * as vscode from "vscode";
+
+export class MDBToCB {
+    static async export(
```
Comment (lokesh-couchbase) [inline src/tools/MDBToCBExport.ts:8]: maybe use migrate instead of export

**Review by lokesh-couchbase — CHANGES_REQUESTED** [review]
Nice Work @prajwal-pai77 , Few more changes and this feature will rock

File: package.json (hunk 1 above)
Comment (prajwal-pai77) [reply package.json]: Changed to mdbMigrate everywhere.

File: src/handlers/handleCLIDownloader.ts (hunk 5 above)
Comment (prajwal-pai77) [reply src/handlers/handleCLIDownloader.ts:385]: Moved all to a seperate function for each.

File: src/handlers/handleCLIDownloader.ts (hunk 6 above)
Comment (prajwal-pai77) [reply src/handlers/handleCLIDownloader.ts]: Updated.

File: src/tools/MDBToCBExport.ts (hunk 7 above)
Comment (prajwal-pai77) [reply src/tools/MDBToCBExport.ts]: Yep, we can add this in the next release if CLI supports it by default.

File: src/webViews/tools/dataMdbMigrate.webview.ts (hunk 13 above)
Comment (prajwal-pai77) [reply src/webViews/tools/dataMdbMigrate.webview.ts]: Removed unwanted CSS

File: src/webViews/tools/dataMdbMigrate.webview.ts (hunk 15 above)
Comment (prajwal-pai77) [reply src/webViews/tools/dataMdbMigrate.webview.ts]: Formatted.

File: src/webViews/tools/dataMdbMigrate.webview.ts (hunk 11 above)
Comment (prajwal-pai77) [reply src/webViews/tools/dataMdbMigrate.webview.ts]: Yep, changed it.

File: src/webViews/tools/dataMdbMigrate.webview.ts (hunk 16 above)
Comment (prajwal-pai77) [reply src/webViews/tools/dataMdbMigrate.webview.ts]: yes, added a new line.

File: src/webViews/tools/dataMdbMigrate.webview.ts (hunk 10 above)
Comment (prajwal-pai77) [reply src/webViews/tools/dataMdbMigrate.webview.ts]: Removed.

File: src/webViews/tools/mdbMigrate.webview.ts [hunk 20]
```diff
@@ -0,0 +1,538 @@
...
+          <meta charset="UTF-8">
+          <meta name="viewport" content="width=device-width, initial-scale=1.0">
+          <title>Data Migrate</title>
+          <link href="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.13/css/select2.min.css" rel="stylesheet">
+          <style>
+            .heading {
+                text-align: center;
+            }
+
+            .advanced-header {
```
Comment (lokesh-couchbase) [inline src/webViews/tools/mdbMigrate.webview.ts:15]: This CSS seems extra, please confirm

**Review by lokesh-couchbase — APPROVED** [review]
Nice Work!!

File: src/webViews/tools/mdbMigrate.webview.ts (hunk 20 above)
Comment (prajwal-pai77) [reply src/webViews/tools/mdbMigrate.webview.ts]: Yes, cleaned it up and pushed.

**Pull request merged**